### ISSUE TITLE
Support for custom notification actions

### ISSUE THREAD

**Issue by MaikuB** [MaikuB/flutter_local_notifications#17]
The plugin should provide the ability to specify custom notification actions. However, this will depends on the Flutter engine being able to support headless-Dart code as per https://github.com/flutter/flutter/issues/6192 and https://github.com/flutter/flutter/issues/3671

It appears Android support is there but will wait to see on if the engine can support it for iOS applications, and for Flutter to provide abstractions to access the functionality. Without the support being added in, then this won't work for scenarios like when the application has been terminated as the logic associated with the action would've been defined in Dart. 

Update with remaining work (IMO)

- [x] rework callbacks to deal with "foreground" and "background" notification actions. See https://github.com/MaikuB/flutter_local_notifications/pull/1489 for more detailed discussion
- [x] potential issue with setting notification categories affecting ability to see the notification actions https://github.com/MaikuB/flutter_local_notifications/issues/1481

Note that last two perhaps could be omitted given market share and those require using deprecated APIs

Edit: 
- there is headless execution on iOS but there are limitations at the moment. See https://github.com/flutter/flutter/issues/23904. Upvote the issue on the Flutter repo if this concerns you to help the Flutter team prioritise
- removed tasks on supporting older iOS and macOS versions given market share and those require using deprecated APIs

**Comment by mit-mit** [comment]
Hi @MaikuB do you have an example use case where of using a custom notification action on iOS?

**Comment by MaikuB** [comment]
@mit-mit 

- a social media app may notify users when someone has replied their post, in which case actions may be presented to allow the user to either reply or like the reply
- similar to the clock app on Android, an app may present a reminder to a user and they can choose to have rescheduled (think snooze)

I believe the support for this would be covered in https://github.com/flutter/flutter/issues/3671 and help enable more fully featured apps. That thread covers other use cases that would require headless-Dart execution

**Comment by mitchhymel** [comment]
@MaikuB 
I worked a bit on my own [notification plugin](https://github.com/mitchhymel/local_notifications) and while it's not fully featured as yours, I did get custom actions working with headless Dart in Android (at least from my testing of using my app daily for the past month, it seems to be working). 

You need to create an [Android service](https://github.com/mitchhymel/local_notifications/blob/master/android/src/main/java/com/mythichelm/localnotifications/services/LocalNotificationsService.java#L32), that will host your shared method channel and support handling the intent in the background. Make sure to set up the shared channel [when the plugin is registered](https://github.com/mitchhymel/local_notifications/blob/master/android/src/main/java/com/mythichelm/localnotifications/LocalNotificationsPlugin.java#L64). Then when you are creating your notification, if it is meant to run in the background, the [intent must be created with your Service class](https://github.com/mitchhymel/local_notifications/blob/master/android/src/main/java/com/mythichelm/localnotifications/entities/NotificationAction.java#L38). Then in your Service's onHandleIntent, you want to [get the shared channel and invoke your callback to dart](https://github.com/mitchhymel/local_notifications/blob/master/android/src/main/java/com/mythichelm/localnotifications/LocalNotificationsPlugin.java#L191). And then [here's](https://github.com/mitchhymel/local_notifications/blob/master/lib/src/local_notifications.dart#L60) where I handle the call from the Android service in Dart.

Hopefully that helps. Great plugin by the way. Your plugin has more features and you seem to be more active in development than I am, so I think I'll add a reference to your plugin in my readme, in case people want the features your plugin offers.

**Comment by MaikuB** [comment]
@mitchhymel thanks for the tips and appreciation :)

However, does that cover when the application has been terminated as well? That's the main scenario i'm looking at it that I have concerns about? If you look at what's in the android_alarm_manager plugin, you can see what they try to do is resolve an instance of the Flutter application etc to find the callback to invoke. That's what I believe to be crucial in handling that scenario and haven't seen in your code. I already handle intents via the plugin as it implements the NewIntentListener

**Comment by mitchhymel** [comment]
I'm not sure if it handles when the application has been terminated. I haven't tested it yet with my code. From your description, and looking at the AlarmService code, I'm thinking my code would not handle the case where the app is terminated. So maybe my code is not a helpful example.

**Comment by MaikuB** [comment]
Fair enough. i've managed to at least handle it so that when app is terminated when comes to tapping the notification and then triggering a callback after launch. Anyway, appreciate you trying to provide help :)

**Comment by MaikuB** [comment]
@mit-mit i mentioned this on the main Flutter repo but don't know if anyone from your team has seen it given there hasn't been a response. I've taken a look at how Dart code is executed from Android and from what I can gather, it doesn't seem like the APIs that have been exposed allow for passing back arguments from the Andoid side back to the Dart function that will be invoked. This prevents being able to implement a lot of use cases e.g. knowing which alarm to snooze once the callback has been received on the Dart side

**Comment by mit-mit** [comment]
@MaikuB let me ask around a bit, but may take a while as the whole team is really busy getting ready for I/O next week.

**Comment by MaikuB** [comment]
Awesome thanks for that. Looking forward to see what's on next week :)

**Comment by MaikuB** [comment]
@mit-mit since I/O is done now, just wanted to see if were you able to find more information about this. I know Hixie said iOS headless execution is being looked into but it'd also be good that the team looks at how to pass payload back from the iOS side as well

**Comment by pinkfish** [comment]
Doesn't the firebase messaging plugin handle this?  I am pretty sure it does processing without pulling the app to the foreground...

**Comment by MaikuB** [comment]
Nope, take a look at the README file for it on how it handles when the app is terminated then you'll see this

> Notification is delivered to system tray. When the user clicks on it to open app onLaunch fires if click_action: FLUTTER_NOTIFICATION_CLICK is set (see below).

**Comment by pinkfish** [comment]
I followed the idea above from @mitchhymel  and it seems to work.  It is opening up the app and calling the method on it to say the button was clicked on.  How do you verify this scenario?  I am swiping the app away in the activity list.  I suspect this will make the startup tricky though since the ordering will likely be an issue when the activity starts.  Since Mitch is using getActivitry on the pending intent it is opening up the app.  Got it working correctly on ios too with categories.

**Comment by pinkfish** [comment]
It won't do background processing in this case though by the looks :)  Which is likely what you are worried about?  It will only work if it forces the app into the foreground and does something.  Which is a bit disappointing.

**Comment by MaikuB** [comment]
What you're doing is actually not that much different to how I handle when the app is terminated and the user taps on the notification, i.e. launch the app and trigger the callback that allows apps to handle the action that occurred. 

Yes, I'm concerned with running code in the background even when the app is terminated. This would help enable scenarios like being able to snooze an alarm and skipping music tracks when the app has been terminated. The mechanisms to achieve this are different (one of which is to have the intent trigger a service) and appear to also handled the case when the app is currently running. However, as I mentioned earlier, the APIs that are available don't allow passing parameters back. I'd much prefer to only add shipping this feature when it's possible to handle all the scenarios (i.e. when is terminated or running). Otherwise, I'd end up with code that needs to be changed later or devs seeing issues raised even when a limitation has been documentation but they didn't bother reading about it, which i'm also guilty of :)

Anyway, hope you understand where I'm coming from

**Comment by MaikuB** [comment]
@pinkfish too right mate :)

**Comment by s-bauer** [comment]
Any Update on this? There are quite a lot of use-cases that require custom actions!

**Comment by MaikuB** [comment]
I'm well aware of that. You should follow the issues linked in the original post. There are open PRs related to this that need to go in first to enable this to work better, particularly when the app isn't running. If it's that urgent and you don't care about scenarios like that via headless Dart execution then you I'd suggest forking to roll a custom build

**Comment by figelwump** [comment]
Hi, with this issue now resolved: https://github.com/flutter/flutter/issues/3671, does that unblock the work for this issue? If so, what are you currently thinking for ETA?

This functionality is critical for a project I'm currently working on, I appreciate your work on this!

**Comment by MaikuB** [comment]
It does unblock from looking at it but it's not rolled over to beta channel and that's where a lot of devs will likely to be on. So if it were done it would be a blocker for release until the changes rolled into beta. Last i checked dev was on 0.5.8 (I believe the changes have rolled into there) and there are reports of hot reloading or restarting being broken (think I ran into an issue too) and devs rolling back to the previous dev build (0.5.7) so there are concerns around stability there. From what I read the reason why beta hasn't been updated for a long time is so the Flutter team can do more rigorous testing as well.

I'm in the midst of the working on an app so hard to give an ETA at the moment though I'll likely get to test the headless execution in the next few days e.g. see if #21 would work on the Dev channel.

If you or someone from the community would be able to assist then that would be great. @pinkfish submitted a PR a while back but that was before headless execution work was done and would be a good starting point

**Comment by figelwump** [comment]
Thanks for the update! I'm curious what you learn from working on #21 (I'll subscribe to that issue as well)

I'm not able to assist at the moment, but I could possibly help in a few weeks. If you or someone else is able to get started in the meanwhile, that would help give this momentum! I'm happy to join in as soon as I can.

**Comment by MaikuB** [comment]
#21 requires implementing all that's required for headless Dart execution to work and the functionality is less complex than doing custom actions so would be easier to test headless Dart execution and setup the foundations for this issue. Not sure how much I'll be able to piece it together but If you followed the issue on the Flutter repo, you'll see that documentation and examples are work in progress. The latter I believe is being done [here](https://github.com/bkonyi/FlutterGeofencing)

**Comment by pinkfish** [comment]
Is this even possible on iOS?  I didn't think you got any callbacks on notifications there.

**Comment by MaikuB** [comment]
Are you referring to #21?

**Comment by pinkfish** [comment]
Right, yes.  Doing a callback on a scheduled notification.  Can you run code then?  You cannot on remote notifications right?

**Comment by MaikuB** [comment]
There is at least this callback, which isn't specific to a scheduled notification when the app is in the foreground

https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter

I should clarify that I wasn't looking at #21 specifically to scheduled notifications but for (local) notifications in general. Headless Dart execution on iOS via the callback I linked isn't necessary but would help to see if it will still works

**Comment by davystrong** [comment]
Any update on this issue? I'd like to use this plugin for an app but the quick actions are essential. I saw a reference to it already being possible on Android, which is the platform I am primarily developing for.

**Comment by MaikuB** [comment]
@davystrong none at the moment. I previously tried to add headless execution of Dart code (this is relevant so code can be run even if the app is terminated) but ran into some issues that appeared to be related to isolates. I had raised an issue in this on the Flutter repo and believe they were going to look at how to make this easier for plugin developers. Am not sure what the status is on this

**Comment by davystrong** [comment]
@MatMaul Are you planning to implement this? I believe this would also solve my problem.

**Comment by vikramaggarwal13** [comment]
This plugin is very supportable in all type of local nofifications but please introduce one more thing where we can pass widget also to notification bar . Where we can show our customized look on nofication area

**Comment by MaikuB** [comment]
@vikramaggarwal13 Based on what I know, I don't think that is achievable. Doing custom layouts for notifications on each platform, requires using the way UI is done on that platform e.g. layout files for Android https://developer.android.com/training/notify-user/custom-notification

**Comment by vikramaggarwal13** [comment]
hmm Okay , But is there any way i want to introduce pause play button on notification area . I want user to play or pause music directly from notification area.

**Comment by MaikuB** [comment]
not at this stage, the reason why have been discussed above already

**Comment by vikramaggarwal13** [comment]
Ok Alright

**Comment by MaikuB** [comment]
To clarify i'm referring to via this plugin. an alternative that you look at writing code that uses platform channels yourself. This may likely require you to invoke the media playback functionality on the Android/iOS side as well

**Comment by martinory** [comment]
Any updates? Not experienced with Android but looked into the code and wonder if it would be possible to do similar to https://github.com/transistorsoft/flutter_background_fetch (which can call Flutter when the application is terminated) and change the HeadlessBroadcastReceiver to listen to respond to notification actions instead?

**Comment by martinory** [comment]
I made it work by changing android_alarm_manager (fork: https://github.com/flutter/plugins/pull/2013) and sending pending intents. This then worked when the app was terminated as well. To create an action I changed the notification details to have 3 actions that contained the text for the action button, method to be invoked and the args as a json string.

```
            Intent actionIntent1 = new Intent(NOTIFICATION_ACTION);
            actionIntent1.putExtra("notificationId", notificationDetails.id);
            actionIntent1.putExtra("method", notificationDetails.firstActionMethod);
            actionIntent1.putExtra("arguments", notificationDetails.firstActionArgs);
            PendingIntent actionPendingIntent1 =
                    PendingIntent.getBroadcast(context.getApplicationContext(), notificationDetails.id + 1, actionIntent1, PendingIntent.FLAG_UPDATE_CURRENT);
            builder.addAction(0, notificationDetails.firstActionText, actionPendingIntent1);

```

**Comment by robindijkhof** [comment]
I read multiple times about "termination". What does this actually mean? Are we talking about "swiping up" from recent apps? Or are we talking about stopping the app from the app-info page in settings?

**Comment by MaikuB** [comment]
@robindijkhof: it means when the app isn't running or has been killed/shutdown through whatever means (e.g. swiping up depending on the OS).

Regarding updating the OP, there's actually a blocked label on this issue already

**Comment by dereklakin** [comment]
@martinory Any chance you could make a PR to add this feature here?

**Comment by MaikuB** [comment]
@dereklakin my understanding of what @martinory has done is dependent on the android alarm manager plugin (would need more details on why there's a dependency on that) so likely only works in specific scenarios. There are other scenarios this should handle where I found it's failing (see previous replies) hence why this is still blocked (note: there are labels on this issue)

**Comment by martinory** [comment]
@MaikuB I did use android_alarm_manager due to that they've already had figured out headless execution of dart code when the app is terminated and that it's well maintained. I did modify the package so it's not the official version of the plugin. In the plugin (android_alarm_manager), there is a todo in the java code that says that they might change the plugin to something similar to the changes I made but I don't know. I did do this just to investigate how it could be solved, there are probably many other ways to do it. Have no idea with iOS because I'm not into that.

I would have a function in my own code similar to:
```
  _channel.setMethodCallHandler((call) {
    final String method = call.method;
    final String args = call.arguments[0];

    switch (method) {
      case "background":
        background();
        break;
      case "triggerAction":
        triggerAction(args);
        break;
      default:
        print('[backgroundCallbackDispatcher] Didn\'t invoke any method ($method)');
        break;
    }
  });
```
And then in the notification details (i also created a copyWith function):
```
notificationDetails.android.copyWith(
              firstActionText: 'TEXT FOR AN ACTION',
              firstActionMethod: 'triggerAction',
              firstActionArgs: actionPayload.serialize(),
```

**Comment by noinskit** [comment]
If anybody's interested, we implemented support for notification actions in a fork (pull request: https://github.com/MaikuB/flutter_local_notifications/pull/338). As @MaikuB wrote, it's not straightforward due to complexity of headless execution, so it's one of multiple possible approaches.

**Comment by MaikuB** [comment]
Thanks for the PR @noinskit. Haven't had a chance to to take a look the PR yet but, I suspect it would still have the same problem with isolates that I reported before as the [issue](https://github.com/flutter/flutter/issues/23904) is still open

**Comment by noinskit** [comment]
@MaikuB I've seen that issue, but I think I'm not hitting it because I'm not using the same platform channel across more than one isolate (or maybe I misunderstood that issue's details).

**Comment by anoopeashok** [comment]
> 
> 
> Fair enough. i've managed to at least handle it so that when app is terminated when comes to tapping the notification and then triggering a callback after launch. Anyway, appreciate you trying to provide help :)

how did you do that ?

**Comment by YannBrrd** [comment]
Hello,

[flutter_background_fecth should](https://github.com/transistorsoft/flutter_background_fetch) do the trick AFAIU. Check [this](https://github.com/transistorsoft/flutter_background_fetch/blob/master/example/lib/main.dart#L71-L73).

**Comment by dluksza** [comment]
Maybe a side note, some time ago I've implemented an Android Home Screen Widget with actions that works when flutter app is terminated.

The way I've achieved that is with dart isolates and `shared_preferences`. When app is starting it sets a shared preference value with dart isolate callback ID. Then when user taps on the Home Screen Widget, intent is emitted and handled on the native side. Intent handler reads isolate callback ID from shared preferences and launches the isolate, adding parameters to detect what action actually was fired from widget.

Then on the Flutter/Dart side, in the isolate handler I'm initializing get_it, recognize user action and execute rest off actions handlers.

This should also work for notification actions. If I find some time I'll crate a PR with POC, but if anyone else would like to try I'm happy to help.

Hope this helps.

**Comment by noinskit** [comment]
@dluksza you're right, this is what https://github.com/MaikuB/flutter_local_notifications/pull/338 is doing as well.

**Comment by dluksza** [comment]
@noinskit that's a pity that #338 was closed. I see the limitations. In my case I use firebase and am able to update data from isolate, then _snapshots_ callback is called on the main app. But can imagine that this won't work for everyone.
One can still use `SendPort` to communicate between isolate and main thread, but this complicate things :|
I do like that you also implemented iOS part (can't say if it works) ;).

**Comment by MaikuB** [comment]
The issue I saw that is logged with the Flutter team (see original post) is that background execution could completely stop working. If you have some way to have it work in the background and main isolate whilst the app is running without such issues then I'd be happy to look at a pull request for it

**Comment by dluksza** [comment]
Well, I would start small. Just provide a way to execute dart code in response to notification action, explicitly document that it runs in separate isolate and can't modify UI unless some kind of external pub-sub service is used (like firestore).

I'm guessing 95% of time main app will not run when notification occurs. Also one can modify local data and start application from isolate using intents.

Having something that can add actions and react to them on dart side is already a huge step forward. It can be later improved, but it is better start with something that works for 70% of use cases, then not have this feature at all.

**Comment by noinskit** [comment]
Yeah, that PR I referred to has limitations and is essentially the "start small" approach. Instead of Firestore is uses shared_preferences (for data) and SendPort (for immediate notifications, if the main part is running).

**Comment by alexander-bourne** [comment]
Please add buttons to notifications, im using your plugin in my app and i need it.

**Comment by bagintz** [comment]
@MaikuB it seems the OneSignal flutter package allows for action buttons and (at least in terms of my very brief test of swiping from recents) it will show notification even if the app is terminated.  Do you know of any reason flutter_local_notifications couldn't do the same?  I am sure it is not that simple but I am wondering what OneSignal might have done to figure it out???

**Comment by MaikuB** [comment]
@bagintz OneSignal is for push notifications not local notifications so it's valid comparison due to different use cases and it works differently as well. From what I've seen, the only thing the buttons will do is launch the app. There was a feature request to allow for running code in the background that I don't believe has been implemented

**Comment by bagintz** [comment]
@MaikuB a feature request in the OneSignal package or flutter_local_notifications?  The buttons only do launch the app but they pass in different values providing for a psuedo multiple action scenario (which is what I need, but I understand that is only one use case).  I was more posting the observation about OneSignal in reply to your Feb 19th post:

> The issue I saw that is logged with the Flutter team (see original post) is that background execution could completely stop working. If you have some way to have it work in the background and main isolate whilst the app is running without such issues then I'd be happy to look at a pull request for it

It would seem that the OneSignal team found some way around that (?)

**Comment by MaikuB** [comment]
> a feature request in the OneSignal package or flutter_local_notifications?

OneSignal

> It would seem that the OneSignal team found some way around that 

I haven't seen anything to indicate that and I'm not sure what you are basing that on

**Comment by bagintz** [comment]
@MaikuB Great, I will look for that on the OneSignal repo.

As for the other I guess maybe I am not sure what the definition of "terminated" is for background execution.  With the Flutter package for OneSignal I can get notifications with multiple action buttons if the app is swiped away, terminated from ADB and upon reboot (with the app not in recents).  Would these cases qualify as "background execution could completely stop working" or is something else I am not thinking of.  I am late to this conversation so I realize I may be missing something.   Thanks!!

**Comment by MaikuB** [comment]
@bagintz yes but buttons should work to invoke code regardless of which state the app is in.

A reminder that an issue has been raised with the Flutter team where they have acknowledged there's are issues/limitations. So if you're doing any digging then I don't think there's much point until there's a solution that the Flutter team can come up with. 

If you have need for things like custom action buttons, upvote on the issue linked in the last paragraph of my original post as it doesn't have many votes compared the reactions I see on this issue about adding notification buttons

**Comment by AresVampire** [comment]
local_notifications SDK provides to add actions to the notification. Can't it be done on flutter_local_notifications? Custom notification actions are widely used in mobile apps.
https://pub.dev/documentation/local_notifications/latest/

**Comment by Mohammad-ezzidden** [comment]
> local_notifications SDK provides to add actions to the notification. Can't it be done on flutter_local_notifications? Custom notification actions are widely used in mobile apps.
> https://pub.dev/documentation/local_notifications/latest/

it is only work if the app not TERMINITED.

**Comment by iofjuupasli** [comment]
Now it's possible to launch app by clicking on notification. Maybe it would be possible to do same for actions? But when opening with action, it will provide some payload to main app?
It won't allow to handle actions in background, without launching app. However for a lot of cases it will be acceptable.
As I understood from this issue, it's probably easier to implement.

**Comment by andzejsw** [comment]
Is it possible to add actions to notification now in the middle of 2020?

**Comment by AbdulMalikDev** [comment]
Been waiting for this for 2 months

**Comment by shashikantdurge** [comment]
This feature was requested 2 yrs ago. Any updates would help us organise better

**Comment by Mohammad-ezzidden** [comment]
disappointing -_-

**Comment by MaikuB** [comment]
@shashikantdurge it wasn't a request that was put in, I put this in myself to track it. There are no more updates as the status is still the same where I mention in the original post mentioned to upvote on the related issue on the Flutter repo that blocks this

**Comment by AbdulMalikDev** [comment]
> @shashikantdurge it wasn't a request that was put in, I put this in myself to track it. There are no more updates as the status is still the same where I mention in the original post to upvote on the related issue on the Flutter repo that blocks this

Is there absolutely ANYTHING we can do other than upvote the issue on the official repo? 

just guide me. Any number to complain or any person to speak to? This has to be done man, its not good if we're ignored for 2 years.

**Comment by MaikuB** [comment]
@AbdulMalikDev You could look at resolving the limitation in the Flutter engine yourself if you can. 

Other thing is you could also looking at writing the code to handle this yourself on the native side. Plugins don't always solve for all cases, in which case developers need to look at writing their own code. For example https://github.com/edsnider/localnotificationsplugin and https://github.com/thudugala/Plugin.LocalNotification are two Xamarin plugins that don't have support for plugins

Similarly, custom notification layouts aren't supported either as those need layouts to be built the standard iOS and Android way

**Comment by ghenry** [comment]
Could you quote me for the work to add this feature @MaikuB if it's possible? 

We sponsored the creation of https://github.com/cloudwebrtc/dart-sip-ua and I'd like action buttons for our app so a user and Accept or Reject and incoming call. And also for messaging for use with https://github.com/vukoye/xmpp_dart

It's best it's added here and open source. 

Thanks (ghenry@surevoip.co.uk)

**Comment by ghenry** [comment]
We're also users of your appauth plugin @MaikuB so it would be good to contribute back to you for that great project.

Thanks!

**Comment by MaikuB** [comment]
@ghenry based on what you've written, sounds like perhaps you haven't been following this issue properly. There are limitations with handling actions using just Dart code when the app isn't running. For your purposes you should look to either (1) fork as the starting point and add in the necessary code on the native side or (2) write the code to handle notifications from scratch

**Comment by ghenry** [comment]
@MaikuB Understood, but based on what you've just written can 1) be abstracted and handled by this plugin on the native side like you are doing?

**Comment by MaikuB** [comment]
@ghenry you could have the notification shown via Dart using a plugin/platform channel but have Java/kotlin code that processes the actions specifically for your application's purpose

**Comment by ghenry** [comment]
> @ghenry you could have the notification shown via Dart using a plugin/platform channel but have Java/kotlin code that processes the actions specifically for your application's purpose

Probably the only way :(

**Comment by AbdulMalikDev** [comment]
https://pub.dev/packages/background_fetch found this package. Would this be helpful to run dart in the background for this use case?

**Comment by matamune94** [comment]
@AbdulMalikDev Yes, background_fetch good working with flutter_local_notification.

**Comment by itssidhere** [comment]
@MaikuB Any updates on this?

**Comment by AbdulMalikDev** [comment]
Any updates?

**Comment by ghenry** [comment]
Maybe we just need a wiki example of how to roll your own? There's some code here for Android:

https://medium.com/@info_67212/firebase-push-notification-with-action-button-in-flutter-a841da348097

and how to get the intents:

https://muetsch.io/how-to-receive-sharing-intents-in-flutter.html
https://flutter.dev/docs/get-started/flutter-for/android-devs#how-do-i-handle-incoming-intents-from-external-applications-in-flutter

The problem is, as @MaikuB explained further up is all this is done in the Android code. Not sure about iOS.

Or am I way off? I have offered to sponsored some type of abstraction -  https://github.com/MaikuB/flutter_local_notifications/issues/17#issuecomment-660651376

**Comment by MaikuB** [comment]
@itssidhere @MaikuB I already mentioned the issue that this is blocked by that you can subscribe to and upvote. If there aren't updates there then you're not going to see updates here.

@ghenry if you need to roll your own then it's a matter of reading the docs for each platform's notification APIs

**Comment by asongkai** [comment]
Any update on this topic? I need to implement a call notification with action button too.

**Comment by ghenry** [comment]
Here's an example on native Android side

https://github.com/doneservices/flutter_callkeep/blob/01e683d16be6d6675cf678d85be683718382a96a/android/src/main/kotlin/co/doneservices/callkeep/CallKeep.kt#L399


https://github.com/doneservices/flutter_callkeep/wiki/Native-android-code-that-handles-incoming-calls

**Comment by asongkai** [comment]
That's not suit the desire requirements.

I only need a notification with action button (decline call or accept call) and time to keep the notification display on top of phone 

I don't need an extra features just only these two buttons and when I click on the actions button it will perform any action only my app and hide the notifications that's it.

**Comment by ghenry** [comment]
Read the code example I pasted then :-)

**Comment by asongkai** [comment]
I just went through it but I still have difficulty on link action between native code and dart code.

**Comment by guyluz11** [comment]
Just to make sure, this include buttons like in the alarm clock app (without opening the message content)?

<img alt="Alarm" height="200" src="https://user-images.githubusercontent.com/9304740/104788547-23606780-579b-11eb-9b35-5c7108a355ab.jpeg">

**Comment by asongkai** [comment]
> Just to make sure, this includes buttons like in the alarm clock app (without opening the message content)?
> 
> <img alt="Alarm" height="200" src="https://user-images.githubusercontent.com/9304740/104788547-23606780-579b-11eb-9b35-5c7108a355ab.jpeg">

Could you share more details please.

**Comment by hanyska** [comment]
Any update?

**Comment by guyluz11** [comment]
> > Just to make sure, this includes buttons like in the alarm clock app (without opening the message content)?
> > <img alt="Alarm" height="200" src="https://user-images.githubusercontent.com/9304740/104788547-23606780-579b-11eb-9b35-5c7108a355ab.jpeg">
> 
> Could you share more details please.

The notification will pop up with two buttons, one to snooze and one for dismiss.
Ringtone will  play in the background in the meantime, and it will stop only when pressing the dismiss button.
Currently turning up the volume or pull from the top to see all notification stops the notification/ringtone sound.
Additionally I need the buttons to signal my app so that it will do staff in the background for each action that was pressed by the buttons.
I am creating open source smart home and I want to link between people alarm clock to actions to execute in there home.
In case the user pressed the snooze button it will not start any action and will "ask" him again later.

**Comment by hanyska** [comment]
> > > Just to make sure, this includes buttons like in the alarm clock app (without opening the message content)?
> > > <img alt="Alarm" height="200" src="https://user-images.githubusercontent.com/9304740/104788547-23606780-579b-11eb-9b35-5c7108a355ab.jpeg">
> > 
> > 
> > Could you share more details please.
> 
> The notification will pop up with two buttons, one to snooze and one for dismiss.
> Ringtone will play in the background in the meantime, and it will stop only when pressing the dismiss button.
> Currently turning up the volume or pull from the top to see all notification stops the notification/ringtone sound.
> Additionally I need the buttons to signal my app so that it will do staff in the background for each action that was pressed.
> I am creating open source smart home and I want to link between people alarm clock to actions to execute in there home.
> In case the user pressed the snooze button it will not start any action and will "ask" him again later.

Can you share notification code example? How did you add the buttons?

**Comment by guyluz11** [comment]
@hanyska it is still not the functionality in my app, this is the functionality that I am trying to understand if the issue will add support for (at list part of it).

**Comment by AbdulMalikDev** [comment]
> @hanyska it is still not the functionality in my app, this is the functionality that I am trying to understand if the issue will add support for (at list part of it).
> The screen shot is example of the same behavior that currently can be found in google alarm clock.

Use awesome_notifications package in flutter. It has better notification capabilities and it is providing action buttons also.

**Comment by guyluz11** [comment]
@AbdulMalikDev What are the benefit and disadvantages of it compared to this library ?

**Comment by AbdulMalikDev** [comment]
> @AbdulMalikDev What are the benefit and disadvantages of it compared to this library ? 

Please check on pub.dev it has a lot of customisation possibilities

**Comment by alistairholmes** [comment]
@MaikuB I was wondering how it's possible for awesome_notifications package to show action buttons for a notification. Have things changed on the Flutter side? Or is that package using some native code behind the scenes?

**Comment by P-B1101** [comment]
Any update?!

**Comment by rafaelsetragni** [comment]
> 
> 
> @MaikuB I was wondering how it's possible for awesome_notifications package to show action buttons for a notification. Have things changed on the Flutter side? Or is that package using some native code behind the scenes?

And im keeping wondering why do you insist in keeping using this plugin if you have awesome_notifications?
If there is some missing feature, i invite everyone to implement it and ill give to you guys all the suport and orientations to develop it.

**Comment by alicankaramil** [comment]
@rafaelsetragni they are literally just asking a question, check your attitude mate. This is not a playground, providing something online doesn't give you the right to act like this.

**Comment by jonas-zebari** [comment]
The plugin awesome_notifications does technically have this functionality, but it is not ready to use for production or any app that is integrated with firebase/cloud messaging. I think having the option to give a local notification an action with a callback makes perfect sense for this package. I am not able to write this code for Android or iOS for lack of knowledge and experience, but I would be happy with work with anyone who might be able to do so.

**Comment by rafaelsetragni** [comment]
In fact, there are already many production mode apps that use awesome_notifications. But as I'm still in the development process, I've chosen not to move the versions to 1.0.0 and beyond because I can make some significant architectural changes along the way, as I will in version 0.0.7.

About `firebase_messaging`, it's not possible to perfectly integrate firebase through the "Flutter layer" with all the `awesome_notifications` features. You only need to use what `firebase_messaging` provides to build the push notifications and `awesome_notifications` to build only local notifications, and also there is some performance issues about that, because dart code execution always involves starting a Flutter engine (sort of a JVM for Dart), which is a heavy task for the device.

I'm also developing a companion plugin to integrate firebase and `awesome_notification` seamlessly into the native layer, with all the awesome_notification features available remotely. This way everything about this integration will be done at the native level, making it possible to achieve maximum performance.

If everything is working fine, this will be the last release of the series in development mode.

**Comment by jonas-zebari** [comment]
I specifically meant that it is not ready for production apps which also integrate with firebase_messaging (as I stated) as having both plugins active causes many issues. I am eagerly waiting for the companion app you mention (though i can't seem to find the repo anymore) and I'm wondering if there is anything I can do to help? I do not know what your plans are for the 0.0.7 release and the companion plugin.

**Comment by rafaelsetragni** [comment]
For now, I just need to finish transcribing the background process for iOS and test what has been done for Android devices. I can estimate the release of version 0.0.7 in two weeks and the complementary plugin in 2 months.

For now, no help is needed to finish these versions, but I'll keep you in touch if I find something that you can help with.

**Comment by hatemragab** [comment]
Any update?

**Comment by MohsenHaydari** [comment]
> For now, I just need to finish transcribing the background process for iOS and test what has been done for Android devices. I can estimate the release of version 0.0.7 in two weeks and the complementary plugin in 2 months.
> 
> For now, no help is needed to finish these versions, but I'll keep you in touch if I find something that you can help with.

:D what's happened?

**Comment by rafaelsetragni** [comment]
> > For now, I just need to finish transcribing the background process for iOS and test what has been done for Android devices. I can estimate the release of version 0.0.7 in two weeks and the complementary plugin in 2 months.
> > For now, no help is needed to finish these versions, but I'll keep you in touch if I find something that you can help with.
> 
> :D what's happened?

That happened:

![image](https://user-images.githubusercontent.com/40064496/147910321-6d5fcb23-9810-4c4a-8c80-c7bb40aa5679.png)

I gonna push Awesome Notifications to the next level, achieving 90%+ covered test cases and automate it with GitHub.

**Comment by MaikuB** [comment]
Thanks to @ened and @Kavantix, the 10.0.0-dev.2 prerelease is now available. This supports actions on Android, iOS and macOS. If the community could use this release to this test, provide feedback and contribute with updates to docs, fixes etc then that would much appreciated.

@proninyaroslav would you by chance be able to help with the Linux side? The way actions are defined are platform-specific. From memory, the original PR you submitted left them out to see if there was going to be a common API for all platforms

**Comment by proninyaroslav** [comment]
@MaikuB 
Ok, I'm on it.

**Comment by MaikuB** [comment]
Thanks @proninyaroslav, much appreciated for this. FYI, I'm doing releases from the 10.0.0 branch so if you plain to make additions, best to do it from there. I merged the PR for actions to master to make it easier to deal with doing other changes. In hindsight, I could've merged it to a different branch but should be fine

**Comment by proninyaroslav** [comment]
@MaikuB 
Done https://github.com/MaikuB/flutter_local_notifications/pull/1442

**Comment by coccyx-cyst** [comment]
@MaikuB Where would the api for actions be seen?

**Comment by alesp** [comment]
I'm testing actions on Android, and NotificationActionDetails sometimes contains an old payload in onSelectNotificationAction. If I add PendingIntent.FLAG_UPDATE_CURRENT to the following code in FlutterLocalNotifications.java:

```
  final PendingIntent actionPendingIntent =
            PendingIntent.getBroadcast(
                context, requestCode++, actionIntent, PendingIntent.FLAG_ONE_SHOT);
```

it properly updates payload every time.

**Comment by Kavantix** [comment]
@alesp  so you use `PendingIntent.FLAG_ONE_SHOT | PendingIntent.FLAG_UPDATE_CURRENT`?

@MaikuB I've also ran into a few issues:

- #1304 happens now when adding notification actions
- The isolate does not seem to start after the user killed the app on android, this used to work when I migrated the code but seems to have broken since

**Comment by MaikuB** [comment]
@Kavantix could it both those issues you mentioned are related since the receiver wouldn't be firing or did you also run into the problem after you added the exported flag on your own fork

**Comment by Kavantix** [comment]
@MaikuB I didnt add the flag yet, I just set the targetSdk to 30 and then the actions worked as long as the app was running, so it might be possible that they are related and are both related to android 12.
However it is still weird that that worked before

**Comment by MaikuB** [comment]
Ok also regarding that payload issue, if I'm not mistaken, that would suggest request codes for notification actions and the notification IDs are clashing. If so, I'm starting to wonder if  request code should be something that is set by as part of using the plugin's APIs. Not ideal but not sure what other approaches are available

**Comment by Kavantix** [comment]
@MaikuB Actually now that you mention this, if they clash that might also influence the starting of the isolate 🤔

**Comment by Kavantix** [comment]
The flag doesnt resolve the second issue, when I click on an action from the example app after killing the app I get this error in logcat:

```
2022-01-22 00:20:28.479 7298-7298/com.dexterous.flutter_local_notifications_example E/cations_exampl: No implementation found for io.flutter.view.FlutterCallbackInformation io.flutter.embedding.engine.FlutterJNI.nativeLookupCallbackInformation(long) (tried Java_io_flutter_embedding_engine_FlutterJNI_nativeLookupCallbackInformation and Java_io_flutter_embedding_engine_FlutterJNI_nativeLookupCallbackInformation__J)
2022-01-22 00:20:28.480 7298-7298/com.dexterous.flutter_local_notifications_example D/AndroidRuntime: Shutting down VM
2022-01-22 00:20:28.480 7298-7298/com.dexterous.flutter_local_notifications_example E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.dexterous.flutter_local_notifications_example, PID: 7298
    java.lang.UnsatisfiedLinkError: No implementation found for io.flutter.view.FlutterCallbackInformation io.flutter.embedding.engine.FlutterJNI.nativeLookupCallbackInformation(long) (tried Java_io_flutter_embedding_engine_FlutterJNI_nativeLookupCallbackInformation and Java_io_flutter_embedding_engine_FlutterJNI_nativeLookupCallbackInformation__J)
        at io.flutter.embedding.engine.FlutterJNI.nativeLookupCallbackInformation(Native Method)
        at io.flutter.view.FlutterCallbackInformation.lookupCallbackInformation(FlutterCallbackInformation.java:28)
        at com.dexterous.flutterlocalnotifications.isolate.IsolatePreferences.lookupDispatcherHandle(IsolatePreferences.java:43)
        at com.dexterous.flutterlocalnotifications.ActionBroadcastReceiver.startEngine(ActionBroadcastReceiver.java:118)
        at com.dexterous.flutterlocalnotifications.ActionBroadcastReceiver.onReceive(ActionBroadcastReceiver.java:84)
        at android.app.ActivityThread.handleReceiver(ActivityThread.java:4348)
        at android.app.ActivityThread.access$1600(ActivityThread.java:256)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2101)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7842)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

**Comment by MaikuB** [comment]
@ened the isolate name server code is something I added that only involves Dart code and is specific to the example app to demonstrate how to have the background isolate communicate with the main one

**Comment by MaikuB** [comment]
@Kavantix the issue you mentioned should be solved now. It happened as the engine is meant to be initialised first before trying to look up the callback handler

**Comment by AdamBridges** [comment]
@MaikuB Do you think it's feasible to expect an option/parameter for an Android notification action to be capable of launching the app / bringing it to the foreground when tapped?

**Comment by MaikuB** [comment]
@AdamBridges I agree it makes sense to achieve the outcome that you mention. What are the issues with `android_intent_plus` and Android 12? I recently addressed the issue with notification actions in Android 12 and I haven't heard of Android 12 issues with the `android_intent_plus` unless I've missed them.

Regarding the intent issue you mention and how it would call the wrong callback, that could be solved by using a different intent action. The plugin can check the action on the intent to determine what it should do. This would be similar to what is done at https://github.com/MaikuB/flutter_local_notifications/blob/f126ddb781d6df88e3c89a93962a123153f107b8/flutter_local_notifications/android/src/main/java/com/dexterous/flutterlocalnotifications/FlutterLocalNotificationsPlugin.java#L1640

This leads me to believe the plugin may require more changes where potentially there are two callbacks when tapping on an action and potentially a breaking change to current behaviour. The resulting approach is similar to how the FCM plugin handles foreground and background messages

1. the current callback in the prerelease gets renamed (e.g. onSelectBackgroundNotificationAction) to indicate that it's only for "background" notification actions as they won't trigger the app to appear
2. this leaves a`onSelectNotificationAction` callback that behaves differently and is only used for "foreground" notification actions where if app is already running, it triggers this callback that runs the main isolate. This would be similar to how the `onSelectNotification` works and allows apps to use this data to update the UI e.g. navigate the user to another page.

Regarding second callback/point, there'd need to be a mechanism to figure out if a notification action launched the app much how [getNotificationAppLaunchDetails](https://pub.dev/documentation/flutter_local_notifications/latest/flutter_local_notifications/AndroidFlutterLocalNotificationsPlugin/getNotificationAppLaunchDetails.html) works. This could be a new method.

Note however, this is something that will require the community to contribute to right now. I can assist in guidance but implementing this is going to need others to step up and assist. It's one of the reasons why there's the "needs help" label here and the plugin already suffers from a lack of contribution from community e.g. a feature request is made with the relevant code changes mentioned in a snippet but author doesn't submit a PR when asked.

One might how argue how much time it's taken to figure things out but one thing that is being neglected is how much time have has been saved by using something that's offer freely?  FYI that right now, this plugin is currently worked on it my spare time and I'm essentially working on something that I don't even use. Therefore I kindly ask that if any of the community here is interested in features etc that they help out to help combat sustainability in open-source.

---
On a semi-related note: FYI that this issue used to be blocked to see if the Flutter team would tackle having a process wide isolate as without it, this would require two callbacks and background isolates use shared preferences. Trying to see if this would be tackled within Flutter was to reduce amount of code that would need to be maintained and avoid potential headaches in trying to figure out a migration path if two callbacks were provided and then the Flutter team implemented a process-wide isolate. However it's been a number of years since that issue was raised with no progress so I think it's fair to say it isn't going to be tackled

**Comment by AdamBridges** [comment]
@MaikuB Okay, I created a PR. It's my first time doing this so hopefully I got it right.

**Comment by rafaelsetragni** [comment]
> @MaikuB Do you think it's feasible to expect an option/parameter for an Android notification action to be capable of launching the app / bringing it to the foreground when tapped? Perhaps a check that issues the same logic that launches the app when a notification is selected?
> 
> Investigated and it certainly is feasible: Looks like a simple parameter check under `FlutterLocalNotificationsPlugin.java` could be implemented for each `actionPendingIntent` value to replace the actionIntent with the default pendingIntent and as an Activity, like so `final PendingIntent actionPendingIntent = PendingIntent.getActivity(context, requestCode++, pendingIntent, actionFlags)`. It's not the best solution since it will likely call whatever method is set to a given notification's onSelectNotification callback, but it's quick and will deliver a solution I'm sure many want. I would try to do this myself but this is new territory for me and it's already taken me a couple hours just to figure this much out.
> 
> Workarounds with `android_intent_plus` is not an option as of Android 12, at least not with `flutter_local_notifications` 10.0.0-dev.6, sadly.

@AdamBridges I did exactly that in Awesome Notification's 0.7.0 beta.

**Comment by rafaelsetragni** [comment]
> @MaikuB Do you think it's feasible to expect an option/parameter for an Android notification action to be capable of launching the app / bringing it to the foreground when tapped? Perhaps a check that issues the same logic that launches the app when a notification is selected?
> 
> Investigated and it certainly is feasible: Looks like a simple parameter check under `FlutterLocalNotificationsPlugin.java` could be implemented for each `actionPendingIntent` value to replace the actionIntent with the default pendingIntent and as an Activity, like so `final PendingIntent actionPendingIntent = PendingIntent.getActivity(context, requestCode++, pendingIntent, actionFlags)`. It's not the best solution since it will likely call whatever method is set to a given notification's onSelectNotification callback, but it's quick and will deliver a solution I'm sure many want. I would try to do this myself but this is new territory for me and it's already taken me a couple hours just to figure this much out.
> 
> Workarounds with `android_intent_plus` is not an option as of Android 12, at least not with `flutter_local_notifications` 10.0.0-dev.6, sadly.

@AdamBridges I did exactly that in Awesome Notification's 0.7.0 beta.

**Comment by MaikuB** [comment]
@noinskit i've updated the original post on what's remaining from my perspective. Link for convenience is https://github.com/MaikuB/flutter_local_notifications/issues/17#issue-316443725. PRs and feedback would be welcome, I remember you did submit one in the past. From memory, dealing with the foreground notification actions should be similar to what you worked on back then. Currently what's in 10.0 invokes logic that runs in the background via isolates for headless execution

**Comment by MaikuB** [comment]
10.0.0-dev.12 has been released where the callbacks have been reworked and changes done around specifying notification categories. This is a breaking change and I have no plans to work on adding support for actions on older versions of iOS and macOS myself so if anyone is interested in support for those OS versions, then they'd need to look at submitting a PR. This may mean this is the last prerelease but I must stress that moving this to a stable release is contingent on the community helping to test, provide feedback and, ideally, contribute fixes

**Comment by AdamBridges** [comment]
Good work, @MaikuB!

Sorry I haven't been able to help yet. I thought I would've had my responsibilities done before the end of March but I underestimated the challenges I had before me. Anyway, I'll try out the updates soon and follow up with any issues and, hopefully, contributions.

**Comment by AdamBridges** [comment]
10.0.0-dev.13 is working well on Android 12 API 31 so far and now that showUserInterface is working as intended, it pretty much solves my originally issue of having an action bring the app to the foreground, so cheers for that. I'll follow up again once I get a chance to test on iOS to share how that works out.

**Comment by MaikuB** [comment]
Thanks for the feedback @AdamBridge and keep me posted. I've not seen more feedback from others though so not sure if that's something to be concerned about...

**Comment by MaikuB** [comment]
@AdamBridges have you had a chance to test to on iOS? From the scenarios covered in the example, I believe this should work for your scenarios too

**Comment by AdamBridges** [comment]
> @AdamBridges have you had a chance to test to on iOS? From the scenarios covered in the example, I believe this should work for your scenarios too

Not yet. Hopefully before the end of the August I'll have a chance and will update then.

**Comment by rich-j** [comment]
@MaikuB we're using this package in our projects and want to start using notification actions. Actions work on Android but crash on iOS.

On a real iPhone Xs running iOS 15.5 we immediately hard crash (i.e. app is terminated) when clicking on the notification action. The Flutter console  shows:

```
*** Assertion failure in -[FlutterEngineManager startEngineIfNeeded:registerPlugins:], FlutterEngineManager.m:73
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'failed to set registerPlugins'
*** First throw call stack:
(0x1db3a9288 0x1f40a3744 0x1dcc36360 0x10077c674 0x1db00ee6c 0x1db010a30 0x1db01ef48 0x1db01eb98 0x1db361800 0x1db31b704 0x1db32ebc8 0x1f7462374 0x1ddc9e648 0x1dda1fd90 0x10052cf78 0x100925ce4)
libc++abi: terminating with uncaught exception of type NSException
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x0000000212de5b38 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`__pthread_kill:
->  0x212de5b38 <+8>:  b.lo   0x212de5b58               ; <+40>
    0x212de5b3c <+12>: pacibsp 
    0x212de5b40 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x212de5b44 <+20>: mov    x29, sp
Target 0: (Runner) stopped.
Lost connection to device.
```

We do see the notification and the action choice. We can successfully click on the notification itself and receive the appropriate callback. It's possible that we are missing something in our initialization. To do further testing, we created a new Flutter demo app using `flutter_local_notifications: ^10.0.0-dev.16` and added the following code to the app:

<details><summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  _initNotification();
  runApp(const MyApp());
}

Future<void> _initNotification() async {
  final FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin = FlutterLocalNotificationsPlugin();

  final AndroidInitializationSettings initializationSettingsAndroid =
      AndroidInitializationSettings('@mipmap/ic_launcher');

  final DarwinInitializationSettings initializationSettingsIOS = DarwinInitializationSettings(
      onDidReceiveLocalNotification: didReceiveLocalNotification,   // Only needed for iOS <10.0 (our minimum)
      notificationCategories: [
        DarwinNotificationCategory("iosActions",
            actions: [DarwinNotificationAction.plain("theAction", "Do the action")])
      ]);

  final InitializationSettings initializationSettings =
      InitializationSettings(android: initializationSettingsAndroid, iOS: initializationSettingsIOS);

  await flutterLocalNotificationsPlugin.initialize(
    initializationSettings,
    onDidReceiveNotificationResponse: didReceiveNotificationResponse,
    onDidReceiveBackgroundNotificationResponse: didReceiveBackgroundNotificationResponse,
  );

  const AndroidNotificationDetails androidPlatformChannelSpecifics = AndroidNotificationDetails(
    'test_channel_id',
    'Test Channel Name',
    channelDescription: 'Notification test channel',
    importance: Importance.max,
    priority: Priority.high,
    actions: [AndroidNotificationAction("theAction", "Do the action")],
  );

  const DarwinNotificationDetails iosPlatformChannelSpecifics = DarwinNotificationDetails(
    subtitle: "subTitleText",
    threadIdentifier: "threadId",
    categoryIdentifier: "iosActions",
  );

  const NotificationDetails platformChannelSpecifics =
      NotificationDetails(android: androidPlatformChannelSpecifics, iOS: iosPlatformChannelSpecifics);

  await flutterLocalNotificationsPlugin
      .show(1, "Action Test", "Notification with action", platformChannelSpecifics, payload: "Payload stuff")
      .catchError((error) => print("Notification failed $error"));
}

//----------
void didReceiveNotificationResponse(NotificationResponse details) =>
    print("Received DidReceiveNotificationResponse ${details.id}, ${details.actionId}, "
        "${details.notificationResponseType}, ${details.payload}");

void didReceiveBackgroundNotificationResponse(NotificationResponse details) =>
    print("Received DidReceiveBackgroundNotificationResponse ${details.id}, ${details.actionId}, "
        "${details.notificationResponseType}, ${details.payload}");

// This callback is only for iOS <10 so we shouldn't need it
void didReceiveLocalNotification(int id, String? title, String? body, String? payload) =>
    print("Received DidReceiveLocalNotification $id, $title");

//----------
class MyApp extends StatelessWidget { ... the rest of the Flutter Demo ...
```

</details>

<details><summary>AppDelegate.swift addition</summary>

```swift
    if #available(iOS 10.0, *) {
      UNUserNotificationCenter.current().delegate = self as? UNUserNotificationCenterDelegate
    }
```

</details>

When we run the above code on Android we're given a notification with action. We can click on the action and receive this in the console:
```
I/flutter (21426): Received DidReceiveBackgroundNotificationResponse 1, theAction, NotificationResponseType.selectedNotificationAction, Payload stuff
```
We can also click on the notification itself (i.e. not the action) and receive:
```
I/flutter (21426): Received DidReceiveNotificationResponse 1, null, NotificationResponseType.selectedNotification, Payload stuff
```

On iOS we can click on the notification and receive (so the notification itself is working):
```
flutter: Received DidReceiveNotificationResponse 1, null, NotificationResponseType.selectedNotification, Payload stuff
```

When we click on the iOS notification action we crash as shown above. We can also run on the iOS simulator which gives us a detailed crash report (over 500 lines) - here's the relevant stack traces:

<details><summary>Crash stack trace synopsis</summary>

```
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Triggered by Thread:  0

Last Exception Backtrace:
0   CoreFoundation                	       0x10e2c75f4 __exceptionPreprocess + 226
1   libobjc.A.dylib               	       0x10e177a45 objc_exception_throw + 48
2   Foundation                    	       0x10ece3874 _userInfoForFileAndLine + 0
3   flutter_local_notifications   	       0x10e083d02 __60-[FlutterEngineManager startEngineIfNeeded:registerPlugins:]_block_invoke + 290 (FlutterEngineManager.m:73)
4   libdispatch.dylib             	       0x10f7228e4 _dispatch_call_block_and_release + 12
5   libdispatch.dylib             	       0x10f723b25 _dispatch_client_callout + 8
6   libdispatch.dylib             	       0x10f731043 _dispatch_main_queue_drain + 1050
7   libdispatch.dylib             	       0x10f730c1b _dispatch_main_queue_callback_4CF + 31
8   CoreFoundation                	       0x10e233ed5 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
9   CoreFoundation                	       0x10e22e6ca __CFRunLoopRun + 2761
10  CoreFoundation                	       0x10e22d704 CFRunLoopRunSpecific + 562
11  GraphicsServices              	       0x114810c8e GSEventRunModal + 139
12  UIKitCore                     	       0x12469265a -[UIApplication _run] + 928
13  UIKitCore                     	       0x1246972b5 UIApplicationMain + 101
14  Runner                        	       0x10ddb222f main + 63 (AppDelegate.swift:5)
15  dyld_sim                      	       0x10dfd2f21 start_sim + 10
16  dyld                          	       0x11ac0051e start + 462

Thread 24:: FlutterLocalNotificationsIsolate.2.ui
0   Flutter                       	       0x1120c312f dart::Utf8::CodeUnitCount(unsigned char const*, long, dart::Utf8::Type*) + 63
1   Flutter                       	       0x1121bc1d3 dart::String::FromUTF8(unsigned char const*, long, dart::Heap::Space) + 35
2   Flutter                       	       0x1123c7f31 dart::kernel::KernelReaderHelper::GetSourceFor(long) + 289
3   Flutter                       	       0x11214d90f dart::kernel::KernelLoader::LoadScriptAt(long, dart::DirectChainedHashMap<dart::kernel::UriToSourceTableTrait>*) + 239
4   Flutter                       	       0x11214c8ce dart::kernel::KernelLoader::InitializeFields(dart::DirectChainedHashMap<dart::kernel::UriToSourceTableTrait>*) + 1758
5   Flutter                       	       0x11214c9ae dart::kernel::KernelLoader::LoadEntireProgram(dart::kernel::Program*, bool) + 94
6   Flutter                       	       0x11243a8df Dart_LoadLibraryFromKernel + 351
7   Flutter                       	       0x1120ac9b2 flutter::DartIsolate::LoadKernel(std::__1::shared_ptr<fml::Mapping const>, bool) + 442
8   Flutter                       	       0x1120acae1 flutter::DartIsolate::PrepareForRunningFromKernel(std::__1::shared_ptr<fml::Mapping const>, bool, bool) + 235
9   Flutter                       	       0x1120b49c4 flutter::KernelIsolateConfiguration::DoPrepareIsolate(flutter::DartIsolate&) + 122
10  Flutter                       	       0x1120aa25a flutter::DartIsolate::CreateRunningRootIsolate(flutter::Settings const&, fml::RefPtr<flutter::DartSnapshot const>, std::__1::unique_ptr<flutter::PlatformConfiguration, std::__1::default_delete<flutter::PlatformConfiguration> >, flutter::DartIsolate::Flags, std::__1::function<void ()>, std::__1::function<void ()> const&, std::__1::function<void ()> const&, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::unique_ptr<flutter::IsolateConfiguration, std::__1::default_delete<flutter::IsolateConfiguration> >, flutter::UIDartState::Context const&, flutter::DartIsolate const*) + 752
11  Flutter                       	       0x1120b68aa flutter::RuntimeController::LaunchRootIsolate(flutter::Settings const&, std::__1::function<void ()>, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::optional<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::unique_ptr<flutter::IsolateConfiguration, std::__1::default_delete<flutter::IsolateConfiguration> >) + 580
12  Flutter                       	       0x111f31c84 flutter::Engine::Run(flutter::RunConfiguration) + 400
13  Flutter                       	       0x111f4996f std::__1::__function::__func<fml::internal::CopyableLambda<flutter::Shell::RunEngine(flutter::RunConfiguration, std::__1::function<void (flutter::Engine::RunStatus)> const&)::$_7>, std::__1::allocator<fml::internal::CopyableLambda<flutter::Shell::RunEngine(flutter::RunConfiguration, std::__1::function<void (flutter::Engine::RunStatus)> const&)::$_7> >, void ()>::operator()() + 119
14  Flutter                       	       0x111e46a1c fml::MessageLoopImpl::FlushTasks(fml::FlushType) + 164
15  Flutter                       	       0x111e4cb98 fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) + 26
16  CoreFoundation                	       0x10e234d6e __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
17  CoreFoundation                	       0x10e23483c __CFRunLoopDoTimer + 915
18  CoreFoundation                	       0x10e233dfd __CFRunLoopDoTimers + 265
19  CoreFoundation                	       0x10e22e3e1 __CFRunLoopRun + 2016
20  CoreFoundation                	       0x10e22d704 CFRunLoopRunSpecific + 562
21  Flutter                       	       0x111e4ccd5 fml::MessageLoopDarwin::Run() + 65
22  Flutter                       	       0x111e4692c fml::MessageLoopImpl::DoRun() + 22
23  Flutter                       	       0x111e4baff void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::function<void (fml::Thread::ThreadConfig const&)> const&, fml::Thread::ThreadConfig const&)::$_0> >(void*) + 187
24  libsystem_pthread.dylib       	    0x7fff701cb4e1 _pthread_start + 125
25  libsystem_pthread.dylib       	    0x7fff701c6f6b thread_start + 15

```

</details>

Please let us know if we're missing any configuration or other processing.

**Comment by Kavantix** [comment]
@rich-j take a look at the readme on the 10.0.0 branch:
https://github.com/MaikuB/flutter_local_notifications/blob/6df12390836fc0fac97badb9227a68fee519e9c6/flutter_local_notifications/README.md?plain=1#L351

**Comment by rich-j** [comment]
@Kavantix thank you for the pointer to the correct `README.md`. I have added the suggested code to my `AppDelegate.swift` and now receive a compile time error of 

```
/IdeaProjects/notification_test/ios/Runner/AppDelegate.swift:11:5: error: cannot find 'FlutterLocalNotificationsPlugin' in scope
        FlutterLocalNotificationsPlugin.setPluginRegistrantCallback { (registry) in
```

Here's my updated `AppDelegate.swift`

```swift
import UIKit
import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    // This is required to make any communication available in the action isolate.
    FlutterLocalNotificationsPlugin.setPluginRegistrantCallback { (registry) in
      GeneratedPluginRegistrant.register(with: registry)
    }
    GeneratedPluginRegistrant.register(with: self)
    // Following line is support for FlutterLocalNotifications
    if #available(iOS 10.0, *) {
      UNUserNotificationCenter.current().delegate = self as? UNUserNotificationCenterDelegate
    }
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}

```

Is there an import needed?

**Comment by Kavantix** [comment]
@rich-j you can import `flutter_local_notifications`

**Comment by rich-j** [comment]
@Kavantix thanks, that works

**Comment by mk-dev-1** [comment]
Hi guys,

lately I am receiving the following crash notification via Crashlytics:

```dart
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: type 'Null' is not a subtype of type 'int'. Error thrown null.
       at MethodChannelFlutterLocalNotificationsPlugin.getNotificationAppLaunchDetails(platform_flutter_local_notifications.dart:65)
       at FlutterLocalNotificationsPlugin.getNotificationAppLaunchDetails(flutter_local_notifications_plugin.dart:199)
       at NotificationManager.init(notification_manager.dart:111)
```

Here's whats happening in NotificationsManager.init():

```dart
    notificationManager = FlutterLocalNotificationsPlugin();
    await notificationManager!.initialize(
      InitializationSettings(
        android: const AndroidInitializationSettings(
          '@mipmap/ic_launcher_foreground',
        ),
        iOS: DarwinInitializationSettings(
          requestSoundPermission: false,
          requestBadgePermission: false,
          requestAlertPermission: false,
          notificationCategories: [
            DarwinNotificationCategory(
              await darwinNotificationCategoryId(NotificationType.rem),
              actions: <DarwinNotificationAction>[
                DarwinNotificationAction.plain(
                    NotificationAction.snooze.toString(),
                    localizations.notificationActionSnooze(snoozeDuration)),
              ],
              options: <DarwinNotificationCategoryOption>{
                DarwinNotificationCategoryOption.allowAnnouncement,
              },
            ),
            DarwinNotificationCategory(
              await darwinNotificationCategoryId(NotificationType.med),
              actions: <DarwinNotificationAction>[
                DarwinNotificationAction.plain(
                    NotificationAction.take.toString(),
                    localizations.notificationActionTake),
                DarwinNotificationAction.plain(
                    NotificationAction.snooze.toString(),
                    localizations.notificationActionSnooze(snoozeDuration)),
              ],
              options: <DarwinNotificationCategoryOption>{
                DarwinNotificationCategoryOption.allowAnnouncement,
              },
            ),
          ],
        ),
      ),
      onDidReceiveNotificationResponse: (response) async {
        if (response.payload != null) {
          handleLocalNotification(response.payload!);
        }
      },
      onDidReceiveBackgroundNotificationResponse:
          onDidReceiveBackgroundNotificationResponse,
    );

    NotificationAppLaunchDetails? launchDetails =
        await notificationManager!.getNotificationAppLaunchDetails(); // <--- This is where the error occurs
    if (launchDetails != null && launchDetails.didNotificationLaunchApp) {
      if (launchDetails.notificationResponse?.payload != null) {
        handleLocalNotification(launchDetails.notificationResponse!.payload!);
      }
    }
```

The method behind that on Android is the following:

```dart
  @override
  Future<NotificationAppLaunchDetails?>
      getNotificationAppLaunchDetails() async {
    final Map<dynamic, dynamic>? result =
        await _channel.invokeMethod('getNotificationAppLaunchDetails');
    final Map<dynamic, dynamic>? notificationResponse =
        result != null && result.containsKey('notificationResponse')
            ? result['notificationResponse']
            : null;
    return result != null
        ? NotificationAppLaunchDetails(
            result['notificationLaunchedApp'],
            notificationResponse: notificationResponse == null
                ? null
                : NotificationResponse(
                    id: notificationResponse['notificationId'],
                    actionId: notificationResponse['actionId'],
                    input: notificationResponse['input'],
                    notificationResponseType: NotificationResponseType.values[ // <--- Error seems to be on this line
                        notificationResponse['notificationResponseType']],
                    payload: notificationResponse.containsKey('payload')
                        ? notificationResponse['payload']
                        : null,
                  ),
          )
        : null;
  }
```

I yet have to find a way to replicate that error myself, however this crash started occurring when I bumped `flutter_local_notifications` from `10.0.0-dev.11` to `10.0.0-dev.14`, and I believe the error was essentially introduced with the changes in `10.0.0-dev.12`.

Is this happening to anyone else? Any ideas?

**Comment by AdamBridges** [comment]
> ```
> if (launchDetails.notificationResponse?.payload != null) {
>         handleLocalNotification(launchDetails.notificationResponse!.payload!);
>       }
> ```

I think the problem may have to do with that line and your use of the Bang (!) operator. Try assigning launchDetails.notificationResponse to a variable and using that to check for null first such as this:

```
NotificationResponse? response = launchDetails.notificationResponse; 
if (response != null && response.payload != null) {
handleLocalNotification(response.payload);
}
```

**Comment by MaikuB** [comment]
@mk-dev-1 is this something you've been able to reproduce yourself? Looks odd that it's happened as the plugin should be returning a `notificationResponseType` from the Android side. There is something I can try out to see if it fixes the issue but be great if you're able to reproduce it so we can see if the change fixes it

**Comment by mk-dev-1** [comment]
@AdamBridges Thank you for looking at it, although I can't see that this should resolve the issue as the error log clearly points to the line before...

@MaikuB Unfortunately I still haven't found a way to reproduce the issue, but it keeps happening in production. From the native Android code in the plugin I also can't really see why that error occurs in the first place. What are you thinking? Maybe I can just incorporate whatever you have in mind in a test version that I distribute to a couple of test users that are affected....

**Comment by MaikuB** [comment]
@mk-dev-1 whilst this should be harmless change, I've done it on a separate branch for now https://github.com/MaikuB/flutter_local_notifications/tree/10.0.0_launchIntent. You should be able to see the commit with the differences. I suspect at some point the `launchIntent` read in different from what gets read to actually get the launch details. If so, then this is something that should've been noticed sooner but I suppose your crash reports make it more noticeable. Let me know how it goes

**Comment by mk-dev-1** [comment]
Thank you so much. I will give it a go with your changes and report back. It will take a couple of days for sure though.

**Comment by AdamBridges** [comment]
> @AdamBridges Thank you for looking at it, although I can't see that this should resolve the issue as the error log clearly points to the line before...

@mk-dev-1 Yeah, I wasn't certain. My logic was that because your logic insists that the `NotificationResponse` will never be null, it's somehow being insisted in the Android `getNotificationAppLaunchDetails()` method and tries to return a `NotificationResponse`, but since the `notificationResponseType` parameter cannot return a Null value (whereas `id`, `payload`, `actionId`, and `input` can), it decides to crash.

**Comment by mk-dev-1** [comment]
I can confirm I no longer see any crashes with the latest changes. Thank you once again for your great work!

**Comment by MaikuB** [comment]
@mk-dev-1 this is now incorporated in the 10.0.0-dev.19 prerelease. Not sure how you referenced the fix in your app but the branch I had mentioned will be deleted soon. As you're using the pre-release and from what I recall, have been using the plugin for quite a while, are you able to share your experience in using the pre-release? I assume you've used it to make use of notification actions. Trying to figure out when to promote it to a stable release though the issue you found would've been one that's existed for quite a while now...

**Comment by noinskit** [comment]
Hi, I'm finally giving this a try, works great so far! Until now I was using my own fork that handled notification actions in my app.

I already have one question:
Is it possible to reinitialize notification categories on iOS? They're now part of `DarwinInitializationSettings` passed to `initialize`. My problem is that in some languages user's gender tends to impact the notification action title.
Or is it safe to call `initialize` multiple times and expect the last configuration to be used?

**Comment by mk-dev-1** [comment]
> @mk-dev-1 this is now incorporated in the 10.0.0-dev.19 prerelease. Not sure how you referenced the fix in your app but the branch I had mentioned will be deleted soon. As you're using the pre-release and from what I recall, have been using the plugin for quite a while, are you able to share your experience in using the pre-release? I assume you've used it to make use of notification actions. Trying to figure out when to promote it to a stable release though the issue you found would've been one that's existed for quite a while now...

After testing it and using it in the beta channel of my app, I have been using the 10.0 prereleases in production for quite a while without any major issues or negative feedback. I don't see any reason to not promote it to a stable release ;-)

**Comment by MaikuB** [comment]
@noinskit you'd need to try to confirm but the underlying native Apple API call is one that sets the categories so would think it should work. This might call for a platform-specific API exposed via the plugin so you don't have to call `initalize()` several times. Happy to take a PR for it though I'm thinking to make this a stable release soon given what @mk-dev-1 has said. Curious though if you managed to get around to testing iOS @AdamBridges?

**Comment by noinskit** [comment]
@MaikuB a new platform-specific API for (re)initializing iOS categories sounds reasonable to me. I agree that it should not block 1.0.0.

**Comment by AdamBridges** [comment]
> @noinskit you'd need to try to confirm but the underlying native Apple API call is one that sets the categories so would think it should work. This might call for a platform-specific API exposed via the plugin so you don't have to call `initalize()` several times. Happy to take a PR for it though I'm thinking to make this a stable release soon given what @mk-dev-1 has said. Curious though if you managed to get around to testing iOS @AdamBridges?

Soon. Probably sometime in September so I'll follow up then.

**Comment by MaikuB** [comment]
Has anyone run into issue #1694? I was going to do a stable release but then saw this reported. It looks like a pub cache issue than a plugin issue though

**Comment by mk-dev-1** [comment]
> Has anyone run into issue #1694? I was going to do a stable release but then saw this reported. It looks like a pub cache issue than a plugin issue though

I believe you are right. Can't say I have seen this issue...

**Comment by MaikuB** [comment]
Forgot to close this out since 10.0 was moved to stable and there's been more updates since then so will do so now. Thanks all for the contribution :)
